pkg/ring/incremental_transfer.go

```diff
@@ -0,0 +1,254 @@
+			err := i.incTransferer.RequestChunkRanges(ctx, ranges, addr, true)
+			if err != nil {
+				level.Error(util.Logger).Log("msg", "failed to request chunks", "target", addr, "ranges", PrintableRanges(ranges), "err", err)
+			}
+
+			completesMtx.Lock()
+			defer completesMtx.Unlock()
+			pendingCompletes[addr] = append(pendingCompletes[addr], ranges...)
+		})
+
+		// Add the token into the ring now.
+		i.addToken(token)
+		r.AddTokens(i.ID, i.getTokens(), false)
+
+		if err := i.updateConsul(ctx); err != nil {
+			level.Error(util.Logger).Log("msg",
+				fmt.Sprintf("failed to update consul when changing token %d to ACTIVE", token))
+		}
+	}
+
+	go func() {
+		if i.cfg.TransferFinishDelay != time.Duration(0) {
+			<-time.After(i.cfg.TransferFinishDelay)
+		}
+
+		completesMtx.Lock()
+		defer completesMtx.Unlock()
+
+		for addr, rgs := range pendingCompletes {
+			i.incTransferer.RequestComplete(context.Background(), rgs, addr)
+		}
+	}()
+
+	return nil
+}
+
+// leaveIncrementalTransfer will attempt to incrementally send chunks to
+// neighboring lifecyclers that should contain data for token ranges the
+// leaving lifecycler will no longer receive writes for.
+func (i *Lifecycler) leaveIncrementalTransfer(ctx context.Context) error {
+	// Make sure all tokens are set to leaving, even when we
+	// fail.
+	defer func() {
+		i.checkRemainingTokens()
+
+		i.setTokens(nil)
+		i.setState(LEAVING)
+		i.updateConsul(ctx)
+	}()
+
+	r := i.getLastRing()
+
+	replicationFactor := i.cfg.RingConfig.ReplicationFactor
+	if active := r.FindIngestersByState(ACTIVE); len(active) <= replicationFactor {
+		return fmt.Errorf("not transferring out; number of ingesters less than or equal to replication factor")
+	}
+
+	success := true
+
+	i.setTransitioningTokens(i.getTokens())
+	tokens := i.getTransitioningTokens()
+
+	for _, token := range tokens {
+		healthy := func(t TokenDesc) bool {
+			// When we're leaving the ring, any of our tokens > our current token
+			// are "healthy" (i.e., they are still in a "ACTIVE" state).
+			if t.Ingester == i.ID {
+				return t.Token > token
+			}
+
+			ing := r.Ingesters[t.Ingester]
+			return ing.IsHealthy(Write, i.cfg.RingConfig.HeartbeatTimeout)
+		}
+
+		workload, ok := i.findTransferWorkload(r, token, healthy)
+		if !ok {
+			success = false
+		}
+
+		workload.Do(func(addr string, ranges []TokenRange) {
+			err := i.incTransferer.SendChunkRanges(ctx, ranges, addr)
+			if err != nil {
+				level.Error(util.Logger).Log("msg", "failed to send chunks", "target_addr", addr, "ranges", PrintableRanges(ranges), "err", err)
+				success = false
+			}
+		})
+
+		// Remove the token from the ring.
+		i.removeToken(token)
+		if err := i.updateConsul(ctx); err != nil {
+			level.Error(util.Logger).Log("msg",
+				fmt.Sprintf("failed to update consul when removing token %d", token))
+			success = false
+		}
+	}
+
+	if !success {
+		return fmt.Errorf("incremental transfer out incomplete")
```

Comment:
Single failure to updateConsul will cause this error, but if the last updateConsul succeeded and everything else was fine, perhaps it can still be considered as a success? Can we reuse call to `checkRemainingTokens` here to check if all is fine?